tempesta_fw/sock.c

```diff
@@ -1065,6 +1068,41 @@ ss_getpeername(struct sock *sk, struct sockaddr *uaddr, int *uaddr_len)
 }
 EXPORT_SYMBOL(ss_getpeername);
 
+static inline void
+__ss_do_send(SsWork *sw)
+{
+	struct sock *sk = sw->sk;
+
+	bh_lock_sock(sk);
+	ss_do_send(sk, &sw->skb_list, sw->flags);
+	if (sw->flags & SS_SEND_F_CONN_CLOSE) {
+		ss_do_close(sk);
+		bh_unlock_sock(sk);
+		SS_CALL(connection_drop, sk);
+		sock_put(sk); /* paired with ss_do_close() */
+		return;
+	}
+	bh_unlock_sock(sk);
+}
+
+static inline void
+__ss_do_close(SsWork *sw)
+{
+	struct sock *sk = sw->sk;
+
+	bh_lock_sock(sk);
+	if (!ss_sock_live(sk)) {
+		SS_DBG("[%d]: %s: Socket inactive: sk %p\n",
+		       smp_processor_id(), __func__, sk);
+		bh_unlock_sock(sk);
+		return;
+	}
+	ss_do_close(sk);
+	bh_unlock_sock(sk);
+	SS_CALL(connection_drop, sk);
+	sock_put(sk); /* paired with ss_do_close() */
```

Comment:
These 4 lines are just copy & paster among this function and `__ss_do_send()`. Also why don't you check `ss_sock_live()` in `__ss_do_send()`? I think it's better to make `__ss_do_close()` as just these 4 lines with the `ss_sock_live()` check (if it's actually needed) and call it directly from `ss_tx_action()` under the socket lock.